### PR TITLE
Reload the checkpoint donor when inference fused the loaded module

### PR DESCRIPTION
**Pull request by raimbekovm**
## summary

- `predict()` and `val()` fuse the loaded module in place, and `Model.train()` hands that same module to the trainer as the pretrained donor, so a fine-tune started after either transferred 108 of 708 tensors and trained from near scratch with only a log line as a trace
- `Model.train()` now reads the donor back from its checkpoint file when a module carries the forward that `fuse()` installs, keeps the facade's architecture and in-memory class names on it, and skips the reload when the run seeds from elsewhere: `pretrained=False`, a `pretrained` path, a resumed run, `reset_weights()`, or weights loaded from an object
- `Model.load()` with a module or state dict clears `pt_path` and the `pretrained` override, so neither `train()` nor a DDP child replaces those weights with the file they came from; `self.ckpt` stays, because its truthiness is what routes the in-memory module to the trainer, and clearing it seeds `load(<module>).train()` from the file instead

## measured

cpu, coco8, one epoch, tensors `BaseModel.load` transfers into the fresh training model:

| sequence | before | after |
| --- | --- | --- |
| `predict()` or `val()`, then `train()`: yolo26n / -seg / -cls / rtdetr-l | 108/708, 128/844, 39/236, 322/941 | 708/708, 844/844, 234/236, 941/941 |
| same, then `train(data=[coco8, coco8])`, python trainer and cli subprocess | 108/708 per dataset | 708/708 per dataset |
| `predict()`, rename `model.model.names`, `train()`: rows remapped by name | 80/80 of 108/708 | 80/80 of 708/708 |
| `YOLO("yolo26n.pt").load("yolo11n.pt")`, `predict()`, `train()`: architecture, weights | yolo26n, 108/708 from neither | yolo26n, 402/708 from yolo11n |
| `reset_weights()`, `load(<module>)`, `resume=True`, `pretrained=False`, `train()` alone | unchanged | unchanged, no extra read |
| `load(<module>)`, `predict()`, `train()`: no file describes the weights | 108/708 | 108/708 |
| `load(<module>)` carrying a zeroed tensor, `train()`: that tensor at `on_train_start` | 0, in-memory weights | 0, in-memory weights; 90.6 from the file once `load()` clears `self.ckpt` |
| `predict()`, `save(p)`, `YOLO(p).train()`: the file itself is fused, tracked separately | 108/708 | 108/708 |
| `YOLO("yolo26n.yaml").load(<module>)`, `train()`: loaded weights reach the trainer, tracked separately | no | no |
| extra checkpoint read, only on a fused module that seeds the run | none | 15 ms for yolo26n |

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

`Model.train()` now reloads an unfused checkpoint donor when inference has fused the in-memory model, preventing subsequent fine-tuning from using the incomplete fused weights.

### 📊 Key Changes

- Detects fused modules during training and reloads their checkpoint before passing the donor to the trainer.
- Preserves the current model architecture configuration and class names on the reloaded donor.
- Applies the same donor-selection logic to multi-dataset training through `MultiTrainer`.
- Skips checkpoint reloading for resumed runs, `pretrained=False`, explicit alternate pretrained paths, reset weights, and in-memory module or state-dict loads.
- `reset_weights()` now disables pretrained initialization, while `load()` clears the file path and pretrained override for in-memory weights.

### 🎯 Purpose & Impact

- Fine-tuning after `predict()` or `val()` can receive the complete checkpoint weights instead of the fused module’s reduced set of transferable tensors.
- In-memory weights remain authoritative when loaded directly, while checkpoint-backed models can be reloaded only when the training setup requires it.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @raimbekovm, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/engine/model.py [hunk 1]
```diff
@@ -335,6 +336,9 @@ def load(self, weights: str | Path = "yolo26n.pt") -> Model:
         if isinstance(weights, (str, Path)):
             self.overrides["pretrained"] = weights  # remember the weights for DDP training
             weights, self.ckpt = load_checkpoint(weights)
+        else:  # in-memory weights have no file to reload from
+            self.model.pt_path = None
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/model.py]: ⚠️ **HIGH**: Loading an in-memory module or state dict does not invalidate the facade's existing checkpoint metadata: `self.ckpt`, `self.ckpt_path`, and `overrides['model']` still reference the previous file. After `YOLO('yolo26n.pt').load(module).predict().train()`, `src` is `None`, but the stale `self.ckpt` makes line 873 pass the fused donor again, so training can still start from only the fused subset. `resume=True` can also resume the old checkpoint, and DDP can reconstruct the old file instead of using the in-memory weights. Clear or replace all stale file metadata and route in-memory donors explicitly.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed the complete 47-line change in `ultralytics/engine/model.py`. The checkpoint reload logic addresses the normal fused-file path, but stale checkpoint/model metadata after in-memory loads leaves realistic single-process, resume, and DDP paths using incorrect weights. One high-severity finding remains.

<details><summary>💬 Posted 1 inline comment</summary>

- ⚠️ **HIGH** `ultralytics/engine/model.py:340` Loading an in-memory module or state dict does not invalidate the facade's existing checkpoint metadata: `self.ckpt`, `self.ckpt_path`, and `overrides['model']` still reference the previous file. After `YOLO('yolo26n.pt').load(module).predict().train()`, `src` is `None`, but the stale `self.ckpt` makes line 873 pass the fused donor again, so training can still start from only the fused subset. `resume=True` can also resume the old checkpoint, and DDP can reconstruct the old file instead of usin…
</details>

File: ultralytics/engine/model.py [hunk 2]
```diff
@@ -848,10 +839,38 @@ def train(
...
+        if (
+            pretrained is True
+            and not args.get("resume")
+            and loaded is not False
+            and any(m.forward == getattr(m, "forward_fuse", None) for m in donor.modules())
+        ):
+            # predict() and val() fuse the loaded module in place, so its tensors can no longer seed training
+            src = loaded if isinstance(loaded, (str, Path)) else getattr(donor, "pt_path", None)
+            if src:
+                donor, _ = load_checkpoint(src)
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/model.py]: 💡 **MEDIUM**: This reload is not sufficient when the source checkpoint is already fused. For example, `model.predict(); model.save('fused.pt'); YOLO('fused.pt').train()` leaves `donor.pt_path == 'fused.pt'`, so this loads the same fused topology and then passes it to `get_model`; unfused BatchNorm and related state are still absent, causing training to start with mismatched or randomly initialized layers. Only reload a known-unfused donor, or preserve the original unfused checkpoint when fusing/saving.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Since the previous review, the donor selection and reset/pretrained handling were added, but the stale checkpoint metadata issue remains unresolved and still affects in-memory loads, resume, and DDP. A saved-already-fused checkpoint is another case where the new reload does not restore trainable weights. Not LGTM.

<details><summary>💬 Posted 2 inline comments</summary>

- ⚠️ **HIGH** `ultralytics/engine/model.py:340` Still unresolved from review 1: this branch clears only `model.pt_path` and `overrides['pretrained']`, but leaves `self.ckpt`, `self.ckpt_path`, and `overrides['model']` pointing to the old checkpoint. With `YOLO('yolo26n.pt').load(module).predict().train()`, `src` is `None`, so `donor` remains fused; the truthy stale `self.ckpt` then enters line 871 and transfers that fused donor instead of the in-memory weights. `resume=True` can also select the old checkpoint, and DDP serializes the stale mo…
- 💡 **MEDIUM** `ultralytics/engine/model.py:854` This reload is not sufficient when the source checkpoint is already fused. For example, `model.predict(); model.save('fused.pt'); YOLO('fused.pt').train()` leaves `donor.pt_path == 'fused.pt'`, so this loads the same fused topology and then passes it to `get_model`; unfused BatchNorm and related state are still absent, causing training to start with mismatched or randomly initialized layers. Only reload a known-unfused donor, or preserve the original unfused checkpoint when fusing/saving.
</details>

File: ultralytics/engine/model.py [hunk 3]
```diff
@@ -335,6 +336,9 @@ def load(self, weights: str | Path = "yolo26n.pt") -> Model:
         if isinstance(weights, (str, Path)):
             self.overrides["pretrained"] = weights  # remember the weights for DDP training
             weights, self.ckpt = load_checkpoint(weights)
+        else:  # no file to reload from; ckpt stays, it routes this module to the trainer
+            self.model.pt_path = None
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/model.py:340]: ⚠️ **HIGH**: This still leaves the facade's checkpoint metadata pointing at the previous file. For `YOLO("yolo26n.pt").load(module)`, `self.ckpt_path` and `overrides["model"]` remain `yolo26n.pt`: `resume=True` therefore resumes that old checkpoint, while multi-GPU training serializes the stale `args.model` and each DDP child reloads the old file instead of the in-memory module. A YAML facade is also broken because its initially empty `self.ckpt` makes line 871 skip the in-memory donor entirely. Route training/DDP from explicit current-module state and invalidate or update the old checkpoint metadata together.

File: ultralytics/engine/model.py (hunk 2 above)
Comment (UltralyticsAssistant) [inline ultralytics/engine/model.py:854]: 💡 **MEDIUM**: The reload source is not checked for being unfused. With `model.predict(); model.save("fused.pt"); YOLO("fused.pt").train()`, `donor.pt_path` is `fused.pt`, so this loads the same fused topology and passes it to `get_model()`. `BaseModel.load()` then intersects the fused state dict with a fresh unfused architecture, leaving BatchNorm and other unfused state absent and reproducing partial transfer. Preserve/use the original unfused checkpoint, or skip this reload when the source checkpoint is already fused.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 3

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The donor selection and reset/pretrained gating now address the ordinary predict/val-then-train path, but both prior correctness issues remain. Stale checkpoint metadata can override in-memory weights in resume and DDP flows, and reloading an already-fused checkpoint does not restore the missing unfused state. Not LGTM.

<details><summary>💬 Posted 2 inline comments</summary>

- ⚠️ **HIGH** `ultralytics/engine/model.py:340` This still leaves the facade's checkpoint metadata pointing at the previous file. For `YOLO("yolo26n.pt").load(module)`, `self.ckpt_path` and `overrides["model"]` remain `yolo26n.pt`: `resume=True` therefore resumes that old checkpoint, while multi-GPU training serializes the stale `args.model` and each DDP child reloads the old file instead of the in-memory module. A YAML facade is also broken because its initially empty `self.ckpt` makes line 871 skip the in-memory donor entirely. Route train…
- 💡 **MEDIUM** `ultralytics/engine/model.py:854` The reload source is not checked for being unfused. With `model.predict(); model.save("fused.pt"); YOLO("fused.pt").train()`, `donor.pt_path` is `fused.pt`, so this loads the same fused topology and passes it to `get_model()`. `BaseModel.load()` then intersects the fused state dict with a fresh unfused architecture, leaving BatchNorm and other unfused state absent and reproducing partial transfer. Preserve/use the original unfused checkpoint, or skip this reload when the source checkpoint is al…
</details>

**Comment by raimbekovm** [comment]
Two review points stay open after three rounds; both are measured identical on `main` and on this branch, and neither has a remedy inside this change.

**`load(<module>)` keeps `ckpt`, `ckpt_path`, `overrides["model"]` on the original file.** `self.ckpt` truthiness is what routes the in-memory module into `trainer.get_model(weights=...)`: with it cleared, a donor carrying a zeroed `model.0.conv.weight` arrives at `on_train_start` with the file's weights (90.58) instead of 0.0. The branch clears the two fields that would otherwise substitute the file for those weights, `pt_path` (reload source) and the `pretrained` override (DDP child), and leaves the rest. `resume=True` after `load()` is the existing contract (#24523: `load()` is name-matched transfer, not a resume entry point), and a DDP child rebuilds from `vars(trainer.args)` on every facade today. Routing the donor on the current module instead is not an option: a yaml facade takes its random init before the trainer's `init_seeds`, so two `YOLO("yolo26n.yaml")` constructions differ while two `train(seed=0, deterministic=True)` runs are identical at `on_train_start` on `main`. The real gap behind that suggestion, `YOLO("yolo26n.yaml").load(<module>).train()` dropping the loaded weights, is pre-existing (same on `main`) and tracked separately.

**A checkpoint that is already fused on disk.** `predict(); save(p); YOLO(p).train()` writes a file with 0 BatchNorm layers and trains 108/708 on `main` and here alike. There is no unfused source to reload; fusion is not reversible, and the owner is `save()`/the in-place fusion, tracked separately.
